protocol.md

```diff
@@ -0,0 +1,748 @@
+on the frame type, some bodies are 0 bytes.
+
+The "call req" frame type has a "ttl" field which is used to specify the
+deadline for this request. The transmitter manages the ttl to account for time
+in transit or time to failover, and send it to receivers only for deadline
+propagation purposes.  This allows receivers to know how much time they have
+remaining to complete this portion of the request. Intermediaries are free to
+make retries with backoff as necessary, as long as they are still within the
+ttl.
+
+Here are some details of the message flow from a service A instance A1, through
+a service router R1, into a service B instance B1, and then back. These are not
+the full details of every single field, but they are intended to illustrate
+some less obvious behaviors.
+
+The flow in this example is: `A1 -> R1 -> B1 -> R1 -> A1`
+
+A1 sends "call req" (type 0x03) to R1:
+
+1. select a message id M1 for this frame, in a predictable sequence for easier
+   debugging. This will be used to match up the response to this request. The
+   scope of M1 exists only between A1 and R1 and only for this connection.
+2. generate a unique traceid which will be propagated to any dependent
+   requests. This should only be done if A1 is the edge-most service in the
+   call chain. The point of this unique traceid is to build a span tree that
+   covers all RPCs supporting a given client-visible transaction, not just
+   tracing of a single logical request through multiple intermediaries.
+3. generate a unique spanid which is used to refer to this specific step in the
+   network
+4. set the ttl to the maximum allowable time for this request. If a response
+   cannot be sent within this time, downstream nodes will cancel the request.
+5. A1 doesn't need any headers, so defaults are used
+6. a "call req" body is generated with service_name "service B", application
+   payload args, and checksum
+
+R1 receives "call req" from A1, sends "call req" (0x03) to B1:
+
+1. select message id M2 for this frame
+2. copy traceid from incoming message to traceid of new message
+3. copy spanid from incoming message to parentid of new message
+4. generate unique spanid
+5. copy ttl from incoming message to ttl of new message
+6. copy service name, args, and csum data to new message
+
+B1 receives "call req" from R1, sends "call res" (0x04) to R1:
+
+1. use message id M2 for this frame
+2. send args from application response and compute csum
+
+R1 receives "call res" from B1, send "call res" (0x03) to A1:
+
+1. match incoming message id M2 with existing "call req"
+2. use message id M1 for new message
+3. copy args/csum from incoming message
+
+A1 receives "call res" from R1:
+
+1. match incoming message M1 with existing "call req"
+2. deliver args to application
+
+## Framing
+
+All frames of all types use this structure:
+
+```
+----------------------------------------------
+| 0-7  | size:2 | type:1 | reserved:1 | id:4 |
```

Comment:
Might just be me, but I find these "tables" easier to read if we add a separating lines of `----`s between the rows.